src/util/psbt/error.rs

```diff
@@ -65,9 +65,9 @@ pub enum Error {
         /// Hash-type
         hash_type: PsbtHash,
         /// Pre-image
-        preimage: Vec<u8>,
+        preimage: Box<[u8]>,
```

Comment:
Since vector keep the length on stack, and bytes on heap, and `Box<[u8]>` ... hmmm ... where does it keep the lenght? What I'm aiming at ... is this doing copy and/or reallocation? Wouldn't be terrible, but just double checking.

Since preimage and hash are fixed size, would be slightly better maybe to have a `Box<[u8; 32]>` or even `Box<SomeHashType>` where `SomeHashType` is a newtype over `[u8;32]`?

Reply:
Vec has 3 usize on the stack: pointer, length and capacity while Box has 2 usize on the stack: pointer and length, thus it saves 8 bytes on 64 bit systems at the cost of losing the ability to be changed. I think `into_boxed_slice` involves an allocation (but not the vice versa `into_vec()` https://doc.rust-lang.org/std/primitive.slice.html#method.into_vec

Preimage is not fixed size, but `hash` field could probably use a proper hash type from `bitcoin_hashes`, removing the need for the `hash_type` field and `PsbtHash` enum, but I am going to cite @sanket1729 since he originally developed the code

Reply:
`Box<[u8]>` is a fat pointer, storing both the pointer and the length (`mem::size_of::<Box<[u8]>> == 16` in 64bit).
The difference between a Vec and a Box<[T]> is that in a box the length == capacity.
because of that, converting a Vec to a Box can potentially do a new allocation + copy + free, because it involves calling `shrink_to_fit` which calls `realloc()` if capacity>len

Reply:
@RCasatta, the reason I implemented it this was to avoid repeated code for each hash type. And, the function `insert_psbt_hash_pair` could be generic instead of a macro as per reviews in #478.  I don't feel too strongly about it, but if other contributors feel that it is worth changing to native hash types from `bitcoin_hashes` I don't mind changing the code and raising another PR.

@dpc, although we use preimages as strict `[u8;32]` downstream in rust-miniscript. The BIP174 specification does not have any constraint on the length of those.

Reply:
Thanks. Just making sure it was considered.